dropshot/src/error.rs

```diff
@@ -316,6 +337,649 @@ impl Error for HttpError {
+            }
+
+            /// Returns the [`http::StatusCode`] corresponding to this error status code.
+            ///
+            /// # Note
+            ///
+            /// This is the same as the `Into<http::StatusCode>` implementation, but
+            /// included as an inherent method because that implementation doesn't
+            /// appear in rustdocs, as well as a way to force the type instead of
+            /// relying on inference.
+            pub fn as_status(&self) -> http::StatusCode {
+                self.0
+            }
+
+            /// Returns the `u16` corresponding to this `error status code.
+            ///
+            /// # Note
+            ///
+            /// This is the same as the `Into<u16>` implementation, but
+            /// included as an inherent method because that implementation doesn't
+            /// appear in rustdocs, as well as a way to force the type instead of
+            /// relying on inference.
+            ///
+            /// This method wraps the [`http::StatusCode::as_u16`] method.
+            pub fn as_u16(&self) -> u16 {
+                self.0.as_u16()
+            }
+
+            /// Returns a `&str` representation of the `StatusCode`
+            ///
+            /// The return value only includes a numerical representation of the
+            /// status code. The canonical reason is not included.
+            ///
+            /// This method wraps the [`http::StatusCode::as_str`] method.
+            pub fn as_str(&self) -> &str {
+                self.0.as_str()
+            }
+
+            /// Get the standardised `reason-phrase` for this status code.
+            ///
+            /// This is mostly here for servers writing responses, but could potentially have application
+            /// at other times.
+            ///
+            /// The reason phrase is defined as being exclusively for human readers. You should avoid
+            /// deriving any meaning from it at all costs.
+            ///
+            /// Bear in mind also that in HTTP/2.0 and HTTP/3.0 the reason phrase is abolished from
+            /// transmission, and so this canonical reason phrase really is the only reason phrase you’ll
+            /// find.
+            ///
+            /// This method wraps the [`http::StatusCode::canonical_reason`] method.
+            pub fn canonical_reason(&self) -> Option<&'static str> {
+                self.0.canonical_reason()
+            }
+        }
+
+        impl fmt::Debug for $T {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                fmt::Debug::fmt(&self.0, f)
+            }
+        }
+
+        /// Formats the status code, *including* the canonical reason.
+        impl fmt::Display for $T {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                fmt::Display::fmt(&self.0, f)
+            }
+        }
+
+        impl PartialEq<u16> for $T {
```

Comment:
The `Eq`/`PartialEq` implementations that compare the type with _other_ types can be derived (at least, without some crate that does that). We are already deriving the `PartialEq<Self>` and `Eq<Self>` implementations: https://github.com/oxidecomputer/dropshot/blob/9d99131ccd47ad138f555ed9037fef1ea5f2146a/dropshot/src/error.rs#L354

Regarding `Deref`, we _could_ do that...I feel like there's sort of an attitude that using `Deref` in that way for, essentially, implementing a kind of inheritance is bad style and that `Deref` should only be used for things that are actually pointers. But, I don't care that much about this particular style point personally, I was just following it because I figured others might. However, implementing `Deref` does mean that methods like `is_informational`, `is_success` , and `is_redirection` can also be called on an `ErrorStatusCode`. Stopping people from calling those methods on this type that isn't _that_ important, since they're not going to cause real problems --- they'll just always return `false`. But it is kind of nonsensical, so not exposing them seems good.

IMO just copying and pasting the method signatures is conceptually simpler than using `Deref` for this, although it involves more code. That's not a strongly-held opinion, though, so I'm very open to being convinced otherwise.

Reply:
cab73b6791b9d7b2dee19ec75d84537f4f235c1a added some documentation about this --- what do you think?